pallets/collator-selection/src/lib.rs

```diff
@@ -291,40 +293,81 @@ pub mod pallet {
 
 	#[pallet::call]
 	impl<T: Config> Pallet<T> {
-		/// Set the list of invulnerable (fixed) collators.
+		/// Set the list of invulnerable (fixed) collators. These collators must do some
+		/// preparation, namely to have registered session keys.
+		///
+		/// The call will remove any accounts that have not registered keys from the set. That is,
+		/// it is non-atomic; the caller accepts all `AccountId`s passed in `new` _individually_ as
+		/// acceptable Invulnerables, and is not proposing a _set_ of new Invulnerables.
+		///
+		/// This call does not maintain mutual exclusivity of `Invulnerables` and `Candidates`. It
+		/// is recommended to use a batch of `add_invulnerable` and `remove_invulnerable` instead.
+		/// A `batch_all` can also be used to enforce atomicity. If any candidates are included in
+		/// `new`, they should be removed with `remove_invulnerable_candidate` after execution.
+		///
+		/// Must be called by the `UpdateOrigin`.
 		#[pallet::call_index(0)]
 		#[pallet::weight(T::WeightInfo::set_invulnerables(new.len() as u32))]
-		pub fn set_invulnerables(
-			origin: OriginFor<T>,
-			new: Vec<T::AccountId>,
-		) -> DispatchResultWithPostInfo {
+		pub fn set_invulnerables(origin: OriginFor<T>, new: Vec<T::AccountId>) -> DispatchResult {
```

Comment:
> an individual set_invulnerable that can be batched at the extrinsic level

This is pretty much `add_invulnerable`.

I waivered about the atomicity of this, but I think non-atomic is better (with the recommendation in the docs that add/remove invulnerable is better for many reasons). The reason is that you could set 20 Invulnerables, and one guy either decides not to set his keys or just forgets (probably makes him a bad choice for invulnerable...), but then this ruins the entire call.

Reply:
> However, perhaps it would be good to emit an event when at least one candidate was not set as invulnerable (or an event per candidate not set)

Done